src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,218 @@
+{
+    typedef SuperCellSize BlockDim;
+
+    DataConnector &dc = Environment<>::get().DataConnector();
+
+    /* load FieldTmp without copy data to host */
+    FieldTmp* fieldTmp = &(dc.getData<FieldTmp > (FieldTmp::getName(), true));
+    /* reset density values to zero */
+    fieldTmp->getGridBuffer().getDeviceBuffer().setValue(FieldTmp::ValueType(0.0));
+
+    /* calculate and add the charge density values from all species in FieldTmp */
+    ForEach<VectorAllSpecies, picongpu::detail::ComputeChargeDensity<bmpl::_1,bmpl::int_<CORE + BORDER> >, MakeIdentifier<bmpl::_1> > computeChargeDensity;
+    computeChargeDensity(forward(fieldTmp), currentStep);
+
+    /* add results of all species that are still in GUARD to next GPUs BORDER */
+    EventTask fieldTmpEvent = fieldTmp->asyncCommunication(__getTransactionEvent());
+    __setTransactionEvent(fieldTmpEvent);
+
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldTmp_coreBorder,
+                 fieldTmp->getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+
+    /* cast libPMacc Buffer to cuSTL Buffer */
+    BOOST_AUTO(fieldE_coreBorder,
+                 dc.getData<FieldE > (FieldE::getName(), true).getGridBuffer().
+                 getDeviceBuffer().cartBuffer().
+                 view(BlockDim::toRT(), -BlockDim::toRT()));
+
+    /* run calculation: fieldTmp = | div E * eps_0 - rho | */
+    using namespace lambda;
+    using namespace PMacc::math::math_functor;
+    typedef picongpu::detail::Div<simDim, typename FieldTmp::ValueType> myDiv;
+    algorithm::kernel::Foreach<BlockDim>()
+        (fieldTmp_coreBorder.zone(), fieldTmp_coreBorder.origin(),
+        cursor::make_NestedCursor(fieldE_coreBorder.origin()),
+            _1 = _abs(expr(myDiv())(_2) * EPS0 - _1));
+
+    /* reduce charge derivation (fieldTmp) to get the maximum value */
+    typename FieldTmp::ValueType maxChargeDiff =
+        algorithm::kernel::Reduce()
+            (fieldTmp_coreBorder.origin(), fieldTmp_coreBorder.zone(), PMacc::nvidia::functors::Max());
+
+    /* reduce again across mpi cluster */
+    container::HostBuffer<typename FieldTmp::ValueType, 1> maxChargeDiff_host(1);
+    *maxChargeDiff_host.origin() = maxChargeDiff;
+
+    PMacc::GridController<simDim>& con = PMacc::Environment<simDim>::get().GridController();
+    using namespace PMacc::math;
+    Size_t<simDim> gpuDim = (Size_t<simDim>)con.getGpuNodes();
+    zone::SphericZone<simDim> zone_allGPUs(gpuDim);
+    PMacc::algorithm::mpi::Reduce<simDim> allGPU_reduce(zone_allGPUs);
+    container::HostBuffer<typename FieldTmp::ValueType, 1> maxChargeDiff_cluster(1);
+    allGPU_reduce(maxChargeDiff_cluster, maxChargeDiff_host, _max(_1, _2));
+
+    if(!allGPU_reduce.root()) return;
+
+    this->output_file << currentStep << " " << (*maxChargeDiff_cluster.origin() * CELL_VOLUME).x()
+        << " " << UNIT_CHARGE << std::endl;
```

Comment:
which rank calls this output? The root! But does it have a valid file handler? Because I do not get  any data output in `chargeConservation.dat`.